File: products/LoanHub/apps/frontend/types/borrower.ts
import type { Person } from "@/types/person";
import type { UserRole } from "@/types/auth";

export type EmploymentStatus =
    | "employed"
    | "unemployed"
    | "self_employed"
    | "student"
    | "pensioner";

export interface Borrower {
    id: string;
    user_id: string;
    employment_status: EmploymentStatus;
    employment_type: string | null;
    employer_name: string | null;
    job_title: string | null;
    employment_start_date: string | null;
    monthly_income: string | number | null;
    net_monthly_income: string | number | null;
    other_monthly_income: string | number;
    other_income_source: string | null;
    salary_date: string | null;
    monthly_living_expenses: string | number;
    monthly_debt_repayments: string | number;
    dependants: number;
    residential_status: string | null;
    years_at_address: number | null;
    has_existing_loans: boolean;
    existing_loan_total: string | number;
    bank_name: string | null;
    account_holder_name: string | null;
    account_last_four: string | null;
    consent_to_share_profile: boolean;
    consent_to_share_documents: boolean;
    consent_to_credit_checks: boolean;
    created_at: string;
    updated_at: string;
    person?: Person | null;
}

export interface CreateBorrowerPayload {
    email: string;
    phone: string;
    password_hash: string;
    first_name: string;
    middle_name: string;
    last_name: string;
    gender: "male" | "female" | "other";
    date_of_birth: string;
    national_id: string;
    passport_number: string;
    marital_status: "single" | "married" | "divorced" | "widowed";
    nationality: string;
    district: string;
    town_or_village: string;
    physical_address: string;
    employment_status: EmploymentStatus;
    employer_name: string;
    job_title: string;
    monthly_income: number;
    salary_date: string;
    has_existing_loans: boolean;
    existing_loan_total: number;
    consent_to_share_profile: boolean;
    consent_to_credit_checks: boolean;
}

export interface UpdateBorrowerPayload {
    employment_status?: EmploymentStatus;
    employment_type?: string | null;
    employer_name?: string | null;
    job_title?: string | null;
    employment_start_date?: string | null;
    monthly_income?: number | null;
    net_monthly_income?: number | null;
    other_monthly_income?: number;
    other_income_source?: string | null;
    salary_date?: string | null;
    monthly_living_expenses?: number;
    monthly_debt_repayments?: number;
    dependants?: number;
    residential_status?: string | null;
    years_at_address?: number | null;
    has_existing_loans?: boolean;
    existing_loan_total?: number;
    bank_name?: string | null;
    account_holder_name?: string | null;
    account_last_four?: string | null;
    consent_to_share_profile?: boolean;
    consent_to_share_documents?: boolean;
    consent_to_credit_checks?: boolean;
}

export interface BorrowerCreateResponse {
    user_id: string;
    person_id: string;
    borrower_id: string;
    email: string | null;
    phone: string;
    role: UserRole;
    first_name: string;
    last_name: string;
    district: string | null;
    created_at: string;
}

File: products/LoanHub/apps/backend/database/schemas/borrower.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from database.models.enums import EmploymentStatus


class BorrowerEvaluationFields(BaseModel):
    employment_type: Optional[str] = Field(default=None, max_length=50)
    employment_start_date: Optional[date] = None
    net_monthly_income: Optional[Decimal] = Field(default=None, ge=0)
    other_monthly_income: Decimal = Field(default=Decimal("0.00"), ge=0)
    other_income_source: Optional[str] = Field(default=None, max_length=200)
    monthly_living_expenses: Decimal = Field(default=Decimal("0.00"), ge=0)
    monthly_debt_repayments: Decimal = Field(default=Decimal("0.00"), ge=0)
    dependants: int = Field(default=0, ge=0, le=50)
    residential_status: Optional[str] = Field(default=None, max_length=40)
    years_at_address: Optional[int] = Field(default=None, ge=0, le=100)
    bank_name: Optional[str] = Field(default=None, max_length=120)
    account_holder_name: Optional[str] = Field(default=None, max_length=200)
    account_last_four: Optional[str] = Field(
        default=None,
        pattern=r"^\d{4}$",
        description="Only the final four account digits are stored.",
    )
    consent_to_share_documents: bool = False


class BorrowerBase(BorrowerEvaluationFields):
    user_id: UUID

    employment_status: EmploymentStatus
    employer_name: Optional[str] = None
    employer_group_id: Optional[UUID] = None
    income_day: Optional[int] = Field(default=None, ge=1, le=31)
    job_title: Optional[str] = None
    monthly_income: Optional[Decimal] = Field(default=None, ge=0)
    salary_date: Optional[str] = None

    has_existing_loans: bool = False
    existing_loan_total: Decimal = Field(default=Decimal("0.00"), ge=0)

    consent_to_share_profile: bool = False
    consent_to_credit_checks: bool = False


class BorrowerCreate(BorrowerBase):
    pass


class BorrowerUpdate(BaseModel):
    employment_status: Optional[EmploymentStatus] = None
    employment_type: Optional[str] = Field(default=None, max_length=50)
    employer_name: Optional[str] = None
    employer_group_id: Optional[UUID] = None
    income_day: Optional[int] = Field(default=None, ge=1, le=31)
    job_title: Optional[str] = None
    employment_start_date: Optional[date] = None
    monthly_income: Optional[Decimal] = Field(default=None, ge=0)
    net_monthly_income: Optional[Decimal] = Field(default=None, ge=0)
    other_monthly_income: Optional[Decimal] = Field(default=None, ge=0)
    other_income_source: Optional[str] = Field(default=None, max_length=200)
    salary_date: Optional[str] = None

    monthly_living_expenses: Optional[Decimal] = Field(default=None, ge=0)
    monthly_debt_repayments: Optional[Decimal] = Field(default=None, ge=0)
    dependants: Optional[int] = Field(default=None, ge=0, le=50)
    residential_status: Optional[str] = Field(default=None, max_length=40)
    years_at_address: Optional[int] = Field(default=None, ge=0, le=100)

    has_existing_loans: Optional[bool] = None
    existing_loan_total: Optional[Decimal] = Field(default=None, ge=0)

    bank_name: Optional[str] = Field(default=None, max_length=120)
    account_holder_name: Optional[str] = Field(default=None, max_length=200)
    account_last_four: Optional[str] = Field(default=None, pattern=r"^\d{4}$")

    consent_to_share_profile: Optional[bool] = None
    consent_to_share_documents: Optional[bool] = None
    consent_to_credit_checks: Optional[bool] = None


class BorrowerRead(BorrowerBase):
    id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True,
    }
